### ISSUE TITLE
Easy to get around

### ISSUE THREAD

**Issue by 0dminnimda** [moscow-python-beer/final-class#120]
```python
from final_class import final


@final
class Example(object):
    ...


class NoProtection:
    def __init_subclass__(cls, *args, **kwargs): pass


class BreakTheDefense(type):
    def __new__(cls, name, bases, classdict):
        bases = NoProtection, *bases
        return type.__new__(cls, name, bases, classdict)


class Error(Example, metaclass=BreakTheDefense):  # Works like clockwork
    ...
```

**Comment by sobolevn** [comment]
Yes, this is true. 

This package was released way before `typing.final` was present.
I don't even use it any more.

But, thanks anyway for showing this sample! It might be useful for someone! 👍

**Comment by 0dminnimda** [comment]
What does `typing.final` have to do with it? `typing.final` only affects static type checking and is ignored when interpreted by python when `final_class.final` affects the actual behavior of the class.

**Comment by sobolevn** [comment]
Yes, this is good enough for me.

**Comment by mauvilsa** [comment]
> I don't even use it any more.

@sobolevn why don't you use it anymore? You no longer have a need or are there any reasons for not using this?

> Easy to get around

@0dminnimda why is it important for you that it shouldn't be easy to get around? For me the most important detail is that if I make a class final, then someone else will get an error when they subclass it. If despite seeing this error they want to subclass it anyway is of less important. It is their problem if they use the class in an unintended way.

**Comment by 0dminnimda** [comment]
@mauvilsa First of all, I just pointed this out, at least it is interesting to know it. Secondly, at the time I was playing with `type` and metaclasses and created a rather difficult to workaround implementation of the final class using pure python, and I was interested in trying to get around this package. Thirdly, cpython can create final classes through the C api (eg. bool), so it is very likely that this mechanism can be implemented as a separate package available through pure python. Anyway, if a package says that it makes the class final, and then it happens that this is not entirely true, I am a little disappointed :)

I can assume that some, probably not very well-designed system, will depend on type checking, and wants to prevent users from getting into its internal mechanism, removing the possibility of inheritance. Although there are much fewer such cases than just need to warn users, though it cannot be excluded. If you want to try to stop as many attempts to bypass this as possible, then it becomes a game to trick the user into thinking that he cannot inherit from this class, and how far you go in this is your choice. This reminds me of "__" versus "_", although both end up giving us the ability to access members, but with a different amount of effort.

I want to say that I agree with your opinion and for most cases this package will go, although for this ideally you want to use `typing.final`, but, of course, not everyone uses static type checking, and we want to warn you as much as possible more users. In the end, there will always be a person who can simply copy the original implementation, or bypass the applied barriers (if it's about type comparison and we need exactly that class).

TL; DR if you want to warn the user - this is more than enough, if you want the functionality of the final class, respectively, you want to suppress as many attempts to inherit (which this package does not implement), then this is a game, how tricky mechanism you can think of to scare the user away with its complexity of workaround (well, or use the C api, it will be impossible to get around this with the help of pure python).

---
I can also try to answer why sobolevn no longer uses this: it seems unlikely that something will break if the user inherits from this type, this is more advice to the user, and it may be enough to warn them about this in the documentation, but what is much more important - checking static typing inside the application, in case something went wrong, so `typing.final` is enough.

**Comment by sobolevn** [comment]
> so typing.final is enough.

👍

**Comment by 0dminnimda** [comment]
In fact, the shorter answer is that this package was used in the same way that `typing.final` is used today, to make sure nothing went wrong. Since `typing.final` was added, there is no need to use this package!

**Comment by mauvilsa** [comment]
Thank you very much for the detailed answer! The reason I ask is not that I need an enforced final class. In theory `typing.final` would be enough for me. However, what I really need is a way to check if a class if final or not. From what I see there is no easy way to check this unless doing something complicated with `ast`. With the `final` of this package at least it can easily be checked which is why I ended up here.

I am still thinking how to go about it. In case you are interested what I am working on is https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_cli.html and https://jsonargparse.readthedocs.io/en/stable/. More specifically if a class is used as a type hint and it is a final class, then it shouldn't behave like a sub-class.

**Comment by 0dminnimda** [comment]
> More specifically if a class is used as a type hint and it is a final class, then it shouldn't behave like a sub-class.

I didn't quite understand what you mean by "behave like a sub-class".

@mauvilsa if you need advice on this .. whatever it is .. then direct me to an issue if there is one. If it does not exist, then why are you working on this and haven’t created it yet? :)

I'm interested to see what's there, maybe I'll throw in some ideas.

**Comment by mauvilsa** [comment]
> > More specifically if a class is used as a type hint and it is a final class, then it shouldn't behave like a sub-class.
> 
> I didn't quite understand what you mean by "behave like a sub-class".

The context to understand what I mean is the first two paragraphs in [lightning_cli](https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_cli.html). Essentially any user defined class automatically becomes configurable via command line arguments and config files. When a class is used as a type hint it means that this parameter accepts an instance of this class or any sub-class. Thus in a config file it is necessary to specify both a class import path and its arguments. However, if a class is final it does not make sense to require specifying the class import path. So my need is being able to detect if a class is final so that in this case only expect the args.

For the moment I am just defining a new `final` decorator, see [jsonargparse/typing.py#L298-L306](https://github.com/omni-us/jsonargparse/blob/f17cd2edb516e3dc44be84fac1d1fc2486076cfe/jsonargparse/typing.py#L298-L306). I would have liked to use python's `typing.final`. But I don't see any simple way to detect if a class is final or not.
